Documenting KUDO tasks

zen-dog opened the pull request:
Fixes #113

zen-dog committed:
Documenting KUDO tasks

alenkacz commented on content/docs/tasks.md:
```diff
@@ -0,0 +1,178 @@
+# Tasks
```
I am kind of wondering if this should not go under 'concepts' if we ever clean that up. I just feel like we have too many stuff on the top level https://kudo.dev/docs/ (on my mac I even have to scroll to see all)

This is nothing against this PR I just wonder what is the ideal docs structure and I was trying to avoid putting more things in the root

alenkacz replied:
or under developing operators? 🤔

zen-dog replied:
> I just feel like we have too many stuff on the top level

Tasks absolutely do not belong on the top-level. But I had to start somewhere 🤷‍♀ I think that all three: `Features`, `Developing Operators`, `Examples` and now `Tasks` belong under the same ☂️ Wdyt about:

```
└── Developing Operators
    ├── Introduction (former developing-operators.md)
    ├── Task (tasks.md)
    ├── Features (features.md)
    └── Examples (examples/README.md)
```

alenkacz replied:
I am not sure about features, because e.g. webhook is more about managing operators then developing them (and I put it under features). Maybe we should have the docs separated by personas actually and features are too generic.

So I think I like what you're proposing, we should have developing operators with tasks and then running operators with webhook? 🤔 

Btw. I am fine with doing this restructure in another PR - up to you :)

zen-dog replied:
Ah, webhooks PR hasn't been merged yet but you put it into features? 🤔 But I like the separation in "developing" and "managing". Though, current `feature.md` content `Controlled Parameter Changes` is more about...developing?

alenkacz replied:
yeah ... I don't know. Also where does cleanup plan go, that could be both... 🤔but probably also developing? but I think that pipes go definitely into developing...

zen-dog replied:
Ok, so I left `Features` at the top for now - not sure where it fits best. But the rest looks like the following:

```
Docs/
...
├── Developing Operators
│   ├── Concepts
│   ├── First Operator (former developing-operators.md)
│   └── Tasks
│   └── Examples
│       ├── Apache Flink
│       ├── Apache Kafka
│       └── Apache ZooKeeper
```
wdyt?

alenkacz replied:
I am not sure about the examples as the content of those pages right now seems to me more like operating, that developing

With concepts I am undecided ... right now they more describe the CRD structure, I am not sure you need that when developing actually, maybe more when operating? I don't know, that page is still ... meh ... to me :) but we could start with putting it there

First Operator I would rename to `Getting started`

But overall much better structure than the current one 👍 :)

alenkacz commented on content/docs/tasks.md:
```diff
@@ -0,0 +1,178 @@
+
+## Apply-Task
+
+An apply-task applies (!) templates to the cluster. Pretty simple. Its `spec.resources` field defines a list of templates that will be applied to the cluster. Given a definition like the one from the [first-operator](developing-operators.md#your-first-kudo-operator):
```
could you explain apply that in k8s terms? I guess in our case it's something like create-or-update right?

alenkacz commented on content/docs/tasks.md:
````diff
@@ -0,0 +1,178 @@
+        - deployment.yaml
+```
+
+a task named `app` will launch a deployment defined in `templates/deployment.yaml`.
````
create a deployment resource? Launch feels like too much, it's actually kubernetes controllers that are "launching" something in my head based on the resources we apply :)

alenkacz commented on content/docs/tasks.md:
````diff
@@ -0,0 +1,178 @@
+        - deployment.yaml
+```
+
+**Note:** deleting non-existing resources is always successful. As of KUDO 0.9.0, it will not wait for the resource to be actually removed e.g. k8s imposes a default graceful termination period of 30 seconds when terminating pods, however, a delete-task will be done before that.
````
maybe link this https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/ and mention that we don't wait for finalizers? (so if there are finalizers, we finish that task when api server accepts our deletion request, not when resource is actually removed from the cluster)

alenkacz commented on content/docs/tasks.md:
````diff
@@ -0,0 +1,178 @@
+          key: indexHtml
+```
+
+a pipe-task spec has two fields: a `pod` spec that will generate our `index.html` and a `pipe` spec that defines a list of files that will be persisted. Here we'll generate a `/tmp/index.html` file that we'll save as a `ConfigMap`. It can be referenced with the key `indexHtml` (more about that below). Pipe-task `spec.pod` field must reference a [core/v1 Pod](https://kubernetes.io/docs/reference/g ...
````
Start with capital "A"?

alenkacz replied:
"It can be referenced with the key" -> "It can be referenced within operator.yaml with the key" ? I was immediately thinking you're talking about kubernetes

alenkacz commented on content/docs/tasks.md:
````diff
@@ -0,0 +1,178 @@
+        name: {{ .Pipes.indexHtml }}
+```
+
+We successfully created a [mount volume from the ConfigMap](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#populate-a-volume-with-data-stored-in-a-configmap). Note the parametrized ` {{ .Pipes.indexHtml } }` configMap name. `.Pipes` keyword allows us to address pipe-artifacts and `indexHtml` is the key we defined in the pipe-task above.
````
`{{ .Pipes.indexHtml } }` -> `{{ .Pipes.indexHtml }}`

zen-dog replied:
Nope 😂 vuepress breaks because of `}}` and escaping them doesn't help. 

P.S. I was wondering if anybody notices and you won! 🎉

alenkacz commented on content/docs/tasks.md:
```diff
@@ -0,0 +1,178 @@
+- File generating Pod has to be side-effect free (meaning side-effects that are observable outside of the container like a 3rd party API call) as the container might be executed multiple times on failure. A `restartPolicy: OnFailure` is used for the pipe-pod
+- Only files <1Mb are applicable to be stored as ConfigMap or Secret. A pipe-task will fail should it try to copy files >1Mb
+- As of KUDO 0.9.0 pipe-artifacts can only be used within the same plan they were generated
+- You can pipeline artifacts by creating them in one pipe-task and mounting in a subsequent one
```
hmm I am not sure what this means, can you elaborate? Oh you mean that you can have multiple pipe tasks each reading input from the previous one?

zen-dog replied:
> Oh you mean that you can have multiple pipe tasks each reading input from the previous one?

exactly. would you phrase it differently?

alenkacz reviewed (approved):
Nice docs 👏 I have only few comments/nits/improvements. Nothing blocking :)

zen-dog committed:
Feedback round one

zen-dog committed:
Improved docs hierarchy

```
Docs/
...
├── Developing Operators
│   ├── Concepts
│   ├── First Operator
│   └── Tasks
│   └── Examples
│       ├── Apache Flink
│       ├── Apache Kafka
│       └── Apache ZooKeeper
```

DanielMSchmidt commented on content/docs/tasks.md:
```diff
@@ -0,0 +1,178 @@
+
+## Overview
+
+A task is the basic building block in the KUDO workflow. Plans, phases, and steps are control structures that execute tasks at the end. You've already come across an Apply-task when developing your [first-operator](first-operator.md#your-first-kudo-operator). KUDO (as of 0.9.0) offers three main task types: `Apply`, `Delete`, `Pipe` and one `Dummy` task which is helpful when debugging and testing ...
```
Saying three types and listing four seems a little odd, would try to rephrase this for more clarity

```suggestion
A task is the basic building block in the KUDO workflow. Plans, phases, and steps are control structures that execute tasks at the end. You've already come across an Apply-task when developing your [first-operator](first-operator.md#your-first-kudo-operator). KUDO (as of 0.9.0) offers three main task types: `Apply`, `Delete`, and `Pipe`. Additionally there is a `Dummy` task which is helpful when debugging and testing your operator. All KUDO tasks are defined in the [operator.yaml](first-operator.md#operatoryaml-file) and must have three fields:
```

DanielMSchmidt commented on content/docs/tasks.md:
```diff
@@ -0,0 +1,178 @@
+
+KUDO will apply all the listed resources and wait for all of them to become healthy. "Health" is dependent on the particular resource: a [deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) must have a defined number of spec instances up and running, a [Job](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/) must finish successfully a ...
+
+**Note:** current implementation will apply all resources in the given order so that e.g. a Pod can mount previously created ConfigMap. However, this is not part of the specification and might change in the future (all resources can be applied concurrently). If you need [happens-before-guarantee](https://en.wikipedia.org/wiki/Happened-before) between your resources, use e.g. multiple serial steps ...
```
```suggestion
**Note:** The current implementation will apply all resources in the given order so that e.g. a Pod can mount previously created ConfigMap. However, this is not part of the specification and might change in the future (all resources can be applied concurrently). If you need [happens-before-guarantee](https://en.wikipedia.org/wiki/Happened-before) between your resources, use e.g. multiple serial steps.
```

DanielMSchmidt commented on content/docs/tasks.md:
````diff
@@ -0,0 +1,178 @@
+        - deployment.yaml
+```
+
+**Note:** deleting non-existing resources is always successful. As of version 0.9.0, KUDO will not wait for the resource to be actually removed and will finish the task when the API server accepts the deletion request. So in case of Pods, k8s imposes a default graceful termination period of 30 seconds, however, a delete-task will be done before that. KUDO 0.9.0 will not wait for [resource finaliz ...
````
I'm wondering why, is there a KEP to link to here?

zen-dog replied:
Why what? not waiting for finalizers? just not implemented yet. Also, not clear if that's semantic that operators need.

DanielMSchmidt commented on content/docs/tasks.md:
````diff
@@ -0,0 +1,178 @@
+          key: indexHtml
+```
+
+Pipe-task spec has two fields: a `pod` spec that will generate our `index.html` and a `pipe` spec that defines a list of files that will be persisted. Here we'll generate a `/tmp/index.html` file that we'll save as a `ConfigMap`. It can be referenced within template resources with the key `indexHtml` (more about that below). Pipe-task `spec.pod` field must reference a [core/v1 Pod](https://kubern ...
````
Is the config map removed afterwards? As config maps keep everything in plain text, is this currently really a good mechanism for secrets like custom certificates?

zen-dog replied:
For certificates use pipe `kind: Secret`

Commit:
Update content/docs/tasks.md

Co-Authored-By: Daniel Schmidt <danielmschmidt92@gmail.com>
